lib/util/exec/docker/index.ts

```diff
@@ -1,49 +1,85 @@
-import { uniq } from 'lodash';
-
 type Opt<T> = T | null | undefined;
 
+type VolumesPair = [string, string];
+export type VolumeOption = Opt<string> | Opt<VolumesPair>;
+
 export interface DockerOptions {
   image: string;
+  tag?: Opt<string>;
   dockerUser?: Opt<string>;
-  volumes?: Opt<Opt<string>[]>;
+  volumes?: Opt<VolumeOption[]>;
   envVars?: Opt<Opt<string>[]>;
   cwd?: Opt<string>;
-  tag?: Opt<string>;
-  preCommands?: Opt<string[]>;
-  postCommands?: Opt<string[]>;
+  preCommands?: Opt<Opt<string>[]>;
+  postCommands?: Opt<Opt<string>[]>;
 }
 
-let dockerUser: string; // Set globally, not configurable per-command
-let localDir: string; // Always used as a mapped volume, also is default working directory if none provided per-command
-let cacheDir: string; // Always used as a mapped volume
+let dockerUser: string | null = null; // Set globally, not configurable per-command
+let localDir: string | null = null; // Always used as a mapped volume, also is default working directory if none provided per-command
+let cacheDir: string | null = null; // Always used as a mapped volume
 
 export function setDockerConfig(config): void {
   dockerUser = config.dockerUser;
   localDir = config.localDir;
   cacheDir = config.cacheDir;
 }
 
+function prepareCommands(commands: Opt<Opt<string>[]>): string[] {
+  const pred = (cmd): boolean => typeof cmd === 'string';
+  return commands ? commands.filter(pred) : [];
+}
+
+function expandVolumeOption(x: Opt<VolumeOption>): VolumesPair | null {
+  if (typeof x === 'string') return [x, x];
+  if (Array.isArray(x) && x.length === 2) {
+    const [from, to] = x;
+    if (typeof from === 'string' && typeof to === 'string') {
+      return [from, to];
+    }
+  }
+  return null;
+}
+
+function volumesEql(x: VolumesPair, y: VolumesPair): boolean {
```

Comment:
We actually need a more sophisticated de-duplication, just in case. Let's assume all volume mappings are A:A as mentioned above. Let's say volumes A and B are given, and A is `/tmp/renovate` and B is `/tmp/renovate/foo`. We want to deduplicate B and use only A, because B.length < A.length && B.startsWith(A).

Reply:
Originally, I did deduplication just for smaller test snapshots. It seems like Docker handles both cases though.

Reply:
It's good enough to keep current exec call snapshots same, so I would leave it as-is or even remove completely.

Reply:
Keep it as-is for now.